### PR TITLE
Add mute and unmute functionality.

### PR DESCRIPTION
**Pull request by microbit-robert**

### CODE REVIEW

File: src/microbithal_js.c [hunk 1]
```diff
@@ -454,13 +454,7 @@ void microbit_hal_audio_select_speaker(bool enable) {
 
 // Input value has range 0-255 inclusive.
 void microbit_hal_audio_set_volume(int value) {
-    /*
-    if (value >= 255) {
-        uBit.audio.setVolume(128);
-    } else {
-        uBit.audio.setVolume(value / 2);
-    }
-    */
```
Comment (microbit-robert) [inline src/microbithal_js.c:463]: Copying this flow doesn't seem to work well for the sim?

File: src/board/audio.ts [hunk 2]
```diff
@@ -18,8 +21,43 @@ export class AudioUI {
       // Match the regular audio rate.
       sampleRate: 7812 * 4,
     });
-    this.default = new BufferedAudio(this.context, defaultAudioCallback);
-    this.speech = new BufferedAudio(this.context, speechAudioCallback);
+    this.programVolume = 128;
+    this.gainNode = this.context.createGain();
+    this.setVolume(this.programVolume, true)
+    this.default = new BufferedAudio(this.context, this.gainNode, defaultAudioCallback);
+    this.speech = new BufferedAudio(this.context,this.gainNode, speechAudioCallback);
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]:
```suggestion
    this.speech = new BufferedAudio(this.context, this.gainNode, speechAudioCallback);
```

Missing space between the commas. Are these files formatting via Prettier on save for you?

File: src/board/audio.ts [hunk 3]
```diff
@@ -18,8 +21,43 @@ export class AudioUI {
...
+
+  /**
+   * Maps device volume range to sensible Web Audio gain output.
+   * @param volume 0 - 255
+   * @returns 0 - 1
+   */
+  private convertDeviceVolumeToGain(volume: number) {
+    if (!volume) {
+      return 0
+    }
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]: What does this case handle? 0/255 is 0.

File: src/board/audio.ts (hunk 2 above)
Comment (microbit-robert) [reply src/board/audio.ts]: Ahh, they are not.

File: src/board/audio.ts [hunk 4]
```diff
@@ -76,7 +114,8 @@ class BufferedAudio {
       buffer,
     });
     source.onended = this.callback;
-    source.connect(this.context.destination);
+    this.gainNode.connect(this.context.destination);
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]: Do we do this twice? Would it make more sense to pass in the pre-connected gain node as the logical destination for the buffered audio?

File: src/microbithal_js.c (hunk 1 above)
Comment (microbit-matt-hillsdon) [reply src/microbithal_js.c:463]: We've put this on the agenda to discuss with Damien.
MakeCode appears to use the full volume range here.

**Review by microbit-matt-hillsdon — APPROVED** [review]
LGTM, @microbit-robert please take another look at the changes we made together and then merge if all's well.

**Pull request merged**